src/OpenTelemetry.Instrumentation.Wcf/Implementation/TelemetryPropagationReader.cs

```diff
@@ -0,0 +1,107 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System;
+using System.Collections.Generic;
+using System.ServiceModel.Channels;
+using OpenTelemetry.Internal;
+
+namespace OpenTelemetry.Instrumentation.Wcf.Implementation;
+
+/// <summary>
+/// Pre-defined PropagationReader callbacks.
+/// </summary>
+internal static class TelemetryPropagationReader
+{
+    private static readonly Func<Message, string, IEnumerable<string>> DefaultReader = Compose(HttpRequestHeaders, SoapMessageHeaders);
+
+    /// <summary>
+    /// Reads the values from the SOAP message headers. If the message is not a SOAP message it returns null.
+    /// </summary>
+    /// <param name="request">The incoming <see cref="Message"/> to read trace context from.</param>
+    /// <param name="name">The header name being requested.</param>
+    /// <returns>An enumerable of all the values for the requested header, or null if the requested header was not present on the request.</returns>
+    public static IEnumerable<string> SoapMessageHeaders(Message request, string name)
+    {
+        Guard.ThrowIfNull(request);
+        Guard.ThrowIfNull(name);
+
+        if (request.Version == MessageVersion.None)
+        {
+            return null;
+        }
+
+        var header = TelemetryMessageHeader.FindHeader(name, request.Headers);
+        return header == null ? null : new[] { header.Value };
+    }
+
+    /// <summary>
+    /// Reads the values from the incoming HTTP request headers. If the message was not made via an HTTP request it returns null.
+    /// </summary>
+    /// <param name="request">The incoming <see cref="Message"/> to read trace context from.</param>
+    /// <param name="name">The header name being requested.</param>
+    /// <returns>An enumerable of all the values for the requested header, or null if the requested header was not present on the request.</returns>
+    public static IEnumerable<string> HttpRequestHeaders(Message request, string name)
+    {
+        Guard.ThrowIfNull(request);
+        Guard.ThrowIfNull(name);
+
+        if (!HttpRequestMessagePropertyWrapper.IsHttpFunctionalityEnabled || !request.Properties.TryGetValue(HttpRequestMessagePropertyWrapper.Name, out var prop))
+        {
+            return null;
+        }
+
+        var value = new HttpRequestMessagePropertyWrapper(prop).Headers[name];
```

Comment:
Why force an allocation here (`new HttpRequestMessagePropertyWrapper`)? Seems like this could all be done statically.

```csharp
static class HttpRequestMessagePropertyHelper // Renamed + made the whole class static
{
   public static WebHeaderCollection? GetHeaders(object? httpRequestMessageProperty)
   {
          EnsureEnabled();

          if (httpRequestMessageProperty == null)
             return null; // Could return new() here, or could return a static empty instance if there is a way to make it readonly (not sure about that). Null is cheap

         return ReflectedValues.HeadersFetcher.Fetch(httpRequestMessageProperty);
   } 

   public static bool TryGetHeaderByName(object? httpRequestMessageProperty, string name, out string? value)
   {
        EnsureEnabled();

        if (httpRequestMessageProperty != null)
        {
              var headers = ReflectedValues.HeadersFetcher.Fetch(httpRequestMessageProperty);
              value = headers[name];
              return value != null;
        }

        value = null;
        return false;
   }
}
```

Reply:
Sounds good...I've updated it to be all static